# Table of Contents

01 Importing Libraries <br /> 
02 Importing Data <br />
03 Cleaning Products.csv (Exercise 4.4) <br />
04 Cleaning orders_wrangled.csv (Task 4.5

# 01 Importing Libraries

In [2]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

# 02 Importing Data

In [3]:
# Creating Basic CSV path
path = r'C:\Users\Rajit\Python\11-2023 Instacart Basket Analysis'

In [4]:
# Import "products.csv" data set into notebook using os library
df_prods = pd.read_csv(os.path.join(path, '02 Data' , 'Original Data', 'products.csv'),index_col = False)
# Import "orders_wrangled.csv" data set into notebook using os library
df_ords = pd.read_csv(os.path.join(path, '02 Data' , 'Prepared Data', 'orders_wrangled.csv'),index_col = False)

# 03 Cleaning Products.CSV (Exercise 4.4)

In [5]:
# Check for Missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
# Creating a dataframe of missing values 
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [7]:
# Creating a new dataframe without the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean.shape

(49677, 5)

In [8]:
# Checking for Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [9]:
# Creating new dataframe with no duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

# 04 Cleaning Orders_Wrangled.CSV (Task 4.5)

In [10]:
# Run the df.describe() function on your df_ords dataframe. 
# Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

df_ords.describe()


Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [11]:
# Function for checking for mixed-type columns
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, ' mixed')
  else: print (col, ' consistent')

C:\Users\Rajit\AppData\Local\Temp\ipykernel_7616\3405543548.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


Unnamed: 0  consistent


C:\Users\Rajit\AppData\Local\Temp\ipykernel_7616\3405543548.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


order_id  consistent


C:\Users\Rajit\AppData\Local\Temp\ipykernel_7616\3405543548.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


user_id  consistent


C:\Users\Rajit\AppData\Local\Temp\ipykernel_7616\3405543548.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


order_number  consistent


C:\Users\Rajit\AppData\Local\Temp\ipykernel_7616\3405543548.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


order_dow  consistent


C:\Users\Rajit\AppData\Local\Temp\ipykernel_7616\3405543548.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


order_hour_of_day  consistent


C:\Users\Rajit\AppData\Local\Temp\ipykernel_7616\3405543548.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


days_since_prior_order  consistent


There is no mixed-type data so no need to make any changes

In [12]:
# Check for Missing Values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

After running the command to check for missing values we found that 206,209 values were missing in the days_since_prior_order column. 

In [ ]:
#Addressing Missing Values

In [17]:
# Creating separate dataframe for missing values 
df_ords_mis = df_ords[df_ords['days_since_prior_order'].isnull() == True]
# Displaying missing values 
df_ords_mis.head(10)


Unnamed: 0  order_id  user_id  order_number  order_dow  order_hour_of_day  \
0            0   2539329        1             1          2                  8   
11          11   2168274        2             1          2                 11   
26          26   1374495        3             1          1                 14   
39          39   3343014        4             1          6                 11   
45          45   2717275        5             1          3                 12   
50          50   2086598        6             1          5                 18   
54          54   2565571        7             1          3                  9   
75          75    600894        8             1          6                  0   
79          79    280530        9             1          1                 17   
83          83   1224907       10             1          2                 14   

    days_since_prior_order  
0                      NaN  
11                     NaN  
26                     NaN  
39                     NaN  
45                     NaN  
50                     NaN  
54                     NaN  
75                     NaN  
79                     NaN  
83                     NaN

Explanation:
We identified that the missing values or NaN, so this is likely because some people haven't made more than one order. (Since it's from the **prior** order). It's best not to make any changes. 

In [19]:
#Checking for Duplicate Values
df_dups = df_ords[df_ords.duplicated()]
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

#Addressing Duplicate Values
No duplicates were found, so no changes need to be made

In [22]:
# Exporting df_prods_clean_no_dups to CSV
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_clean.csv'))
#Exporting df_ords (since no changes were made)
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'order_clean.csv'))